File: googleSpeedRun/180425/squaresOfASortedArray.js
/* Given an integer array nums sorted in non-decreasing order, return an array of
 the squares of each number sorted in non-decreasing order.

Example 1:

Input: nums = [-4,-1,0,3,10]
Output: [0,1,9,16,100]
Explanation: After squaring, the array becomes [16,1,0,9,100].
After sorting, it becomes [0,1,9,16,100].

Example 2:

Input: nums = [-7,-3,2,3,11]
Output: [4,9,9,49,121] */

const sortedQuares = nums => {
    let result = [];
    let left = 0;
    let right = nums.length-1;
    for(let i = nums.length-1; i >= 0; i--){
        let leftSquared = Math.pow(nums[left], 2);
        let rightSquared = Math.pow(nums[right], 2);
        if(leftSquared > rightSquared){
            result[i] = leftSquared;
            left++;
        }else{
            result[i] = rightSquared;
            right--;
        }
    }
    return result;
};

//iterando desde nums.length-1 por que sabemos que el ultimo va a ser el mas grandeson :) de atras pa adelante
//en teoria pudiese hacerlo como siempre pero al final lo tendria que hacer un tipo de .reverse o una madre asi xd

console.log(sortedQuares([-4,-1,0,3,10])); //[0,1,9,16,100]

File: python/roadmap/two pointers/squaresOfASortedArray.py
"""
 Given an integer array nums sorted in non-decreasing order, return an 
array of the squares of each number sorted in non-decreasing order.

Example 1:
Input: nums = [-4,-1,0,3,10]
Output: [0,1,9,16,100]
Explanation: After squaring, the array becomes [16,1,0,9,100].
After sorting, it becomes [0,1,9,16,100].

Example 2:
Input: nums = [-7,-3,2,3,11]
Output: [4,9,9,49,121] 
"""

from typing import List
def sortedSquares(nums:List[int]) -> List[int]:
    
    left, right = 0, len(nums) - 1
    ans = []

    while left <= right:
        if abs(nums[left]) > abs(nums[right]):
            ans.append(nums[left] ** 2)
            left += 1
        else:
            ans.append(nums[right] ** 2)
            right -= 1
        
    ans.reverse()
    return ans
                
    

#print(sortedSquares([-4,-1,0,3,10]))

def sortedSquares2(nums:List[int]) -> List[int]:
    
    n = len(nums)
    ans = [0] * n
    left, right = 0, n - 1
    write = n - 1

    while left <= right:
        left_sq = nums[left] ** 2
        right_sq = nums[right] ** 2

        if left_sq > right_sq:
            ans[write] = left_sq
            left += 1
        else:
            ans[write] = right_sq
            right -= 1

        write -= 1

    return ans

print(sortedSquares2([-4,-1,0,3,10]))